File: python/src/webex_message_handler/mention_parser.py
"""Parse Webex ``<spark-mention>`` tags from decrypted HTML."""

from __future__ import annotations

import re
from dataclasses import dataclass, field


@dataclass
class ParsedMentions:
    """Mentions extracted from message HTML."""

    mentioned_people: list[str] = field(default_factory=list)
    mentioned_groups: list[str] = field(default_factory=list)


# Match the whole opening tag with a single bounded ``[^>]*`` (linear), then
# pull attributes out of the captured tag. A prior form used two ``[^>]*``
# around the attribute, which is a polynomial-ReDoS risk under Python's
# backtracking ``re`` engine on crafted ``<spark-mention…`` input.
_MENTION_RE = re.compile(r"<spark-mention[^>]*>", re.IGNORECASE)
_OBJECT_TYPE_RE = re.compile(r'data-object-type="([^"]*)"', re.IGNORECASE)
_PERSON_ID_RE = re.compile(r'data-object-id="([^"]*)"', re.IGNORECASE)
_GROUP_TYPE_RE = re.compile(r'data-group-type="([^"]*)"', re.IGNORECASE)


def parse_mentions(html: str | None) -> ParsedMentions:
    """Extract mentioned people and groups from decrypted HTML.

    Parses ``<spark-mention>`` tags to find person UUIDs and group
    mention types (e.g. ``"all"``). Duplicates are removed.

    Args:
        html: Decrypted HTML content from a Webex message.

    Returns:
        ParsedMentions with ``mentioned_people`` and ``mentioned_groups``.
    """
    result = ParsedMentions()
    if not html:
        return result

    seen: set[str] = set()

    for match in _MENTION_RE.finditer(html):
        tag = match.group(0)
        type_match = _OBJECT_TYPE_RE.search(tag)
        object_type = type_match.group(1) if type_match else ""

        if object_type == "person":
            id_match = _PERSON_ID_RE.search(tag)
            if id_match and id_match.group(1) and id_match.group(1) not in seen:
                seen.add(id_match.group(1))
                result.mentioned_people.append(id_match.group(1))
        elif object_type == "groupMention":
            group_match = _GROUP_TYPE_RE.search(tag)
            if group_match and group_match.group(1):
                key = f"group:{group_match.group(1)}"
                if key not in seen:
                    seen.add(key)
                    result.mentioned_groups.append(group_match.group(1))

    return result

File: go/mention_parser.go
package webexmessagehandler

import "regexp"

// ParsedMentions holds person and group mentions extracted from message HTML.
type ParsedMentions struct {
	MentionedPeople []string
	MentionedGroups []string
}

var (
	mentionRe   = regexp.MustCompile(`(?i)<spark-mention[^>]*data-object-type="([^"]*)"[^>]*>`)
	personIDRe  = regexp.MustCompile(`(?i)data-object-id="([^"]*)"`)
	groupTypeRe = regexp.MustCompile(`(?i)data-group-type="([^"]*)"`)
)

// ParseMentions extracts mentioned people and groups from decrypted HTML.
//
// Parses <spark-mention> tags to find person UUIDs and group mention types
// (e.g. "all"). Duplicates are removed.
func ParseMentions(html string) ParsedMentions {
	result := ParsedMentions{}
	if html == "" {
		return result
	}

	seen := make(map[string]bool)

	matches := mentionRe.FindAllStringSubmatch(html, -1)
	allTags := mentionRe.FindAllString(html, -1)

	for i, match := range matches {
		objectType := match[1]
		tag := allTags[i]

		if objectType == "person" {
			idMatch := personIDRe.FindStringSubmatch(tag)
			if len(idMatch) > 1 && idMatch[1] != "" && !seen[idMatch[1]] {
				seen[idMatch[1]] = true
				result.MentionedPeople = append(result.MentionedPeople, idMatch[1])
			}
		} else if objectType == "groupMention" {
			groupMatch := groupTypeRe.FindStringSubmatch(tag)
			if len(groupMatch) > 1 && groupMatch[1] != "" {
				key := "group:" + groupMatch[1]
				if !seen[key] {
					seen[key] = true
					result.MentionedGroups = append(result.MentionedGroups, groupMatch[1])
				}
			}
		}
	}

	return result
}
